casper/src/main/scala/coop/rchain/casper/MultiParentCasperImpl.scala

```diff
@@ -70,113 +56,210 @@ class MultiParentCasperImpl[F[_]: Sync: Concurrent: Log: Time: SafetyOracle: Las
   private[this] val CreateBlockMetricsSource =
     Metrics.Source(CasperMetricsSource, "create-block")
 
-  def getGenesis: F[BlockMessage] = genesis.pure[F]
+  def getVersion: F[Long] = version.pure[F]
+
+  //TODO rename to getApprovedBlock
+  def getGenesis: F[BlockMessage] = approvedBlock.pure[F]
 
   def getValidator: F[Option[PublicKey]] = validatorId.map(_.publicKey).pure[F]
 
-  def addBlock(b: BlockMessage): F[ValidBlockProcessing] = {
+  // Later this should replace addBlock, but for now one more method created, or there are too many tests to fix
+  def addBlockFromStore(
+      blockHash: BlockHash,
+      allowAddFromBuffer: Boolean
+  ): F[ValidBlockProcessing] = {
+    def returnSafetyRangeNotFilled: F[ValidBlockProcessing] =
+      BlockStatus
+        .exception(
+          new Exception("Not enough blocks past ApprovedBlock received.")
+        )
+        .asLeft[ValidBlock]
+        .pure[F]
 
-    def logAlreadyProcessed: F[ValidBlockProcessing] =
-      Log[F]
-        .info(
-          s"Block ${PrettyPrinter.buildString(b.blockHash)} has already been processed by another thread."
+    for {
+      blockAvailable <- BlockStore[F].get(blockHash)
+      returnNoBlockInStore = BlockStatus
+        .exception(
+          new Exception("Casper tries to add block that is not present in BlockStore")
         )
-        .as(BlockStatus.processing.asLeft)
+        .asLeft[ValidBlock]
+        .pure[F]
+
+      result <- blockAvailable match {
+                 case None =>
+                   returnNoBlockInStore
+                 case Some(b) =>
+                   for {
+                     dag                   <- blockDag
+                     missingDepCheckResult <- Validate.missingBlocks(b, dag)
+                     dagMissingBlockDependencies = missingDepCheckResult.equals(
+                       Left(InvalidBlock.MissingBlocks)
+                     )
+                     isReadyApprovedBlockChild = (isApprovedBlockChild(b) &&^ safetyRangeFilled)
+                     blockIsReady              <- ~^(dagMissingBlockDependencies.pure[F]) ||^ isReadyApprovedBlockChild
+
+                     _ <- fetchMissingDependencies(b).unlessA(!dagMissingBlockDependencies)
```

Comment:
This is like negation on both sides, in the function (unless) and in the argument. _When dagMissingBlockDependencies_ sounds more natural.